src/Stack/Build/Source.hs

```diff
@@ -171,13 +166,9 @@ depPackageHashableContent DepPackage {..} = do
                         then ""
                         else "-" <> encodeUtf8 (T.pack $ C.unFlagName f)
                 flags = map flagToBs $ Map.toList (cpFlags dpCommon)
-                locationTreeKey (PLIHackage _ _ tk) = tk
-                locationTreeKey (PLIArchive _ pm) = pmTreeKey pm
-                locationTreeKey (PLIRepo _ pm) = pmTreeKey pm
-                treeKeyToBs (TreeKey (BlobKey sha _)) = SHA256.toHexBytes sha
                 ghcOptions = map encodeUtf8 (cpGhcOptions dpCommon)
                 haddocks = if cpHaddocks dpCommon then "haddocks" else ""
-                hash = treeKeyToBs $ locationTreeKey pli
+                hash = immutableLocShaBs pli
             return $ B.concat ([hash, haddocks] ++ flags ++ ghcOptions)
```

Comment:
Possible optimization: instead of using `ByteString`s here, use `Builder`s, which will more efficiently concatenate together. May not be worth the overhead of changing things.